### PR TITLE
Add integer atomic types

### PR DESCRIPTION
**Pull request by Amanieu**
Tracking issue: #32976
RFC: rust-lang/rfcs#1543

The changes to AtomicBool in the RFC are not included, they are in a separate PR (#32365).

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks @Amanieu! Some thoughts of mine:

* Can you make sure the `#[cfg(target_has_atomic)]` starts out as unstable as well? The gating here should be similar to things like `target_vendor`.
* Would it be possible to define `Atomic{I,U}size` as part of the same `atomic_int!` macro?

File: src/librustc_back/target/aarch64_apple_ios.rs [hunk 1]
```diff
@@ -23,6 +23,7 @@ pub fn target() -> Target {
         options: TargetOptions {
             features: "+neon,+fp-armv8,+cyclone".to_string(),
             eliminate_frame_pointer: false,
+            max_atomic_width: 128,
```
Comment (alexcrichton) [inline src/librustc_back/target/aarch64_apple_ios.rs]: Just confirming, but this is inherent to aarch64 (e.g. 128-bit atomics are available everywhere?)

File: src/librustc_back/target/x86_64_apple_ios.rs [hunk 2]
```diff
@@ -20,6 +20,9 @@ pub fn target() -> Target {
         target_os: "ios".to_string(),
         target_env: "".to_string(),
         target_vendor: "apple".to_string(),
-        options: opts(Arch::X86_64)
+        options: TargetOptions {
+                max_atomic_width: 64,
```
Comment (alexcrichton) [inline src/librustc_back/target/x86_64_apple_ios.rs]: Indentation looks a little extra here

**Comment by alexcrichton** [comment]
Gah, I hit enter too quickly...

* Would it be possible to detect cmpxchg16b for on x86_64 for more CPUs? Something like querying LLVM? It seems like that'd be a major use case for this, but it's only supported on OSX right now.
* To confirm, platforms like MIPS only have 32-bit atomics but we're saying they have 8-bit atomics, right? Are we really sure that these don't lower to calls in compiler-rt?

I wonder if this can also try to have a more comprehensive test? We don't run tests for platforms like ARM/MIPS/etc, but it'd be good to exercise those platforms as they're quite relevant here.

Also, cc @rust-lang/libs, @brson

File: src/librustc_back/target/aarch64_apple_ios.rs (hunk 1 above)
Comment (Amanieu) [inline src/librustc_back/target/aarch64_apple_ios.rs]: Yes

**Comment by Amanieu** [comment]
> Would it be possible to define Atomic{I,U}size as part of the same atomic_int! macro?

The reason I left them out is because they use different stability attributes (`rust1`, `extended_compare_and_swap` and `atomic_debug`), whereas all the new atomic types are under the `integer_atomics` feature.

> Would it be possible to detect cmpxchg16b for on x86_64 for more CPUs? Something like querying LLVM? It seems like that'd be a major use case for this, but it's only supported on OSX right now.

This is a bit complicated because some early AMD x86_64 CPUs don't support cmpxchg16b. Support for this instruction can be enabled in LLVM by targeting a CPU that supports it (which the OSX target does with `core2`), or by enabling the `cx16` LLVM feature.

I don't think there is a way to extract the maximum atomic size from LLVM's API. I've mostly copied what Clang does, which is to determine it from the target architecture and any compilation flags passed in (`-march`, `-mcx16`, etc).

> To confirm, platforms like MIPS only have 32-bit atomics but we're saying they have 8-bit atomics, right? Are we really sure that these don't lower to calls in compiler-rt?

Yes, these will never be lowered to compiler-rt calls. The test should be able to catch any cases that are lowered to libcalls.

> I wonder if this can also try to have a more comprehensive test? We don't run tests for platforms like ARM/MIPS/etc, but it'd be good to exercise those platforms as they're quite relevant here.

How would you suggest doing this?

**Comment by Amanieu** [comment]
> Can you make sure the #[cfg(target_has_atomic)] starts out as unstable as well? The gating here should be similar to things like target_vendor.

Should I reuse the `integer_atomics` feature gate for this or add a new `cfg_target_has_atomic` feature gate?

**Comment by alexcrichton** [comment]
> The reason I left them out is because they use different stability attributes

Right yeah, but perhaps that could be an input to the macro? (e.g. the macro takes a stability attribute as input)

> I've mostly copied what Clang does

Only statically, though, right? If I pass `-C target-cpu=native` this won't enable cmpxchg16b?

> The test should be able to catch any cases that are lowered to libcalls.

Unfortunately this test won't catch anything because it's only run on x86_64 and x86 for now.

> How would you suggest doing this?

A "minimal cross-compiled libcore" could probably suffice. Something with a few lang items enough to just generate an object file should be enough. You could then call the intrinsics directly to ensure they're not lowered to libcalls.

**Comment by Amanieu** [comment]
> Only statically, though, right? If I pass -C target-cpu=native this won't enable cmpxchg16b?

No, for that we would need to be able to recognize all the CPU variant and set the appropriate LLVM flags for them. See Clang: http://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20130930/090167.html

**Comment by Amanieu** [comment]
Should we expose `compare_and_swap` on the new atomic types? The plan is to deprecate it in favor of `compare_exchange`.

**Comment by Amanieu** [comment]
Also, should I add constants like `ATOMIC_U8_INIT` for all the new integer types?

**Comment by Amanieu** [comment]
> Right yeah, but perhaps that could be an input to the macro? (e.g. the macro takes a stability attribute as input)

Done

> A "minimal cross-compiled libcore" could probably suffice. Something with a few lang items enough to just generate an object file should be enough. You could then call the intrinsics directly to ensure they're not lowered to libcalls.

Done

**Comment by alexcrichton** [comment]
> No, for that we would need to be able to recognize all the CPU variant and set the appropriate LLVM flags for them. See Clang: http://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20130930/090167.html

Yeah that's kinda unfortunate for now, although we're not even exposing the 128-bit types just yet, so it's ok to punt on this for now.

> Should we expose compare_and_swap on the new atomic types?

Nah I wouldn't bother. If it's easier to write the macro where they all have `compare_and_swap` (as isize/usize must have it) then that's also fine (no skin off our back).

> Also, should I add constants like ATOMIC_U8_INIT for all the new integer types?

Indeed!

---

Awesome, thanks for the updates!

**Comment by Amanieu** [comment]
I've added the constants, is there anything else to do?

**Comment by alexcrichton** [comment]
Looks good to me, thanks @Amanieu! I want to double check with @rust-lang/libs about the `#[cfg(target_has_atomic = "ptr")]` for bool/isize/usize/ptr, but I suspect we'll be ok with it for now.

**Comment by ollie27** [comment]
Are the constants not going to be deprecated once `const fn` is stabilized?

**Comment by alexcrichton** [comment]
They likely will be, yes, but `const fn` isn't currently stable so some stable alternative is needed

File: src/doc/reference.md [hunk 3]
```diff
@@ -2079,6 +2079,8 @@ The following configurations must be defined by the implementation:
 * `target_family = "..."` - Operating system family of the target, e. g.
   `"unix"` or `"windows"`. The value of this configuration option is defined
   as a configuration itself, like `unix` or `windows`.
+* `target_has_atomic = "..."` - Set of integer sizes on which the target can perform
+  atomic operations. Values are `"8"`, `"16"`, `"32"`, `"64"` and `"ptr`".
```
Comment (huonw) [inline src/doc/reference.md]: Is `ptr` the word we want to use here? AFAIK, it isn't used anywhere else in the language (although there are examples in `std`) when pointers are discussed e.g. `target_pointer_width`, `isize`/`usize`, `*const`/`*mut`.

**Comment by ranma42** [comment]
> > Only statically, though, right? If I pass -C target-cpu=native this won't enable cmpxchg16b?
>
> No, for that we would need to be able to recognize all the CPU variant and set the appropriate LLVM flags for them.

This could be fixed by #31709, if we added features for the appropriate flags (`cx16` is the one for `cmpxchg16b`).

**Comment by Amanieu** [comment]
@huonw I picked `ptr` somewhat arbitrarily, I don't mind changing it to something else.

@ranma42 That looks great! I wonder if there's a way to get the maximum atomic size for the current target directly. In any case this can probably wait since we don't support 128-bit integers yet.

**Comment by ranma42** [comment]
The atomic lowering code (see [here](http://reviews.llvm.org/D18200)) has a `getMaxAtomicSizeInBitsSupported` function, but right now it would not be a good option for Rust because:
 - it is not available in any LLVM release
 - currently it only works for SPARC

**Comment by Amanieu** [comment]
@ranma42 Yes, ideally we should use that when it becomes available. Which is why I am somewhat uneasy about stabilizing `max_atomic_width` as a target spec option if we're just going to be removing it later anyways. Is it possible to make such an option nightly-only?

**Comment by alexcrichton** [comment]
Nah custom target specs are relatively unstable today, so as long as we have a reasonable default (e.g. target_pointer_width) it probably won't be that widely used and the unstable-ness of target specs will take care of it.

**Comment by Amanieu** [comment]
I set the default to 0, which disables all atomic types.

**Comment by alexcrichton** [comment]
Hm does libcore even compile without atomic types enabled? Surely libstd doesn't compile... In that sense you'd basically be *required* to specify `max_atomic_width`, so perhaps it could by default be the pointer size?

**Comment by Amanieu** [comment]
Yes, libcore compiles fine with atomic types disabled. However none of the other standard crates (liballoc, libstd, etc) will compile due to missing atomic types.

I changed the default to use the target pointer width.

**Comment by alexcrichton** [comment]
The libs team discussed this during triage today (sorry for the delay!) and we were all on board. Thanks again for the PR @Amanieu!

@bors: r+ 60bf93f754c74cbc98d457a7b913e44b7a22bd3b

**Comment by arielb1** [comment]
```

../src/libsyntax/feature_gate.rs:272:5: 272:6 error: no rules expected the token `(`
../src/libsyntax/feature_gate.rs:272     (active, cfg_target_has_atomic, "1.10.0", Some(32976)),
                                         ^
make: *** [x86_64-apple-darwin/stage0/lib/rustlib/x86_64-apple-darwin/lib/stamp.syntax] Error 101
make: *** Waiting for unfinished jobs....
```

**Comment by Amanieu** [comment]
Fixed

**Comment by Amanieu** [comment]
Sorry about that, fixed again.

**Comment by Amanieu** [comment]
Rebased and changed the test to only run on Linux.

**Pull request merged**